### PR TITLE
deprecate quagga 'vtysh' service

### PR DESCRIPTION
**Pull request by gsomlo**
The two patches in this pull request address a timing/synchronization/busy-waiting problem with Quagga's "vtysh" configuration service. The service is deprecated/removed, and each Quagga routing service becomes responsible for the configuration of its own respective daemon.

See the thread entitled "busy-wait / timeout error in quaggaboot.sh" from December 2016 on the core-users mailing list, and the individual commit messages for the full details.

Thanks,
--Gabriel

### CODE REVIEW

File: daemon/core/services/quagga.py [hunk 1]
```diff
@@ -194,52 +178,36 @@ def generateQuaggaBoot(cls, node, services):
...
-    if grep -E -q '^[[:space:]]*router[[:space:]]+pim6?[[:space:]]*$' $QUAGGA_CONF; then
-        vtyfiles="$vtyfiles xpimd.vty"
-    fi
-
-    # wait for Quagga daemon vty files to appear before invoking vtysh
-    waitforvtyfiles $vtyfiles
-
-    $QUAGGA_BIN_DIR/vtysh -b
+    # (re-)run 'vtysh -b' to configure new daemon from integrated config file:
+    $FLOCK_BIN_DIR/flock $QUAGGA_STATE_DIR $QUAGGA_BIN_DIR/vtysh -b
```
Comment (siliconja) [inline daemon/core/services/quagga.py]: Is it possible that this command will wait forever?

From flock(1): 

>        By default, if the
>        lock cannot be immediately acquired, flock  waits  until  the  lock  is
>        available.

It seems like this is not a possibility. **vtysh** would need to hang, and then the script would get stuck anyway.

**Comment by siliconja** [comment]
@tomgoff do you want to review this since you have been deeply involved with both Quagga and CORE?   Thanks!

**Comment by gsomlo** [comment]
The more I think about it, the more I'd like to wrap flock around the whole quaggaboot.sh script call, directly from quagga.py; something like:

<     _startup = ("sh quaggaboot.sh bgpd",)
---
>     _startup = ("flock quaggaboot.sh sh quaggaboot.sh bgpd",)

That way, the entire sequence of the starting daemon and its corresponding "vtysh -b" call are atomic w.r.t. any other daemon's sequence; Also, we could then maybe check and only run vtysh -b when the *last* daemon is started, e.g. by comparing $(ls /var/run/quagga/*.vty | wc -l) against $(grep "^router " $QUAGGA_CONF | wc -l), [1] and only running "vtysh -b" when they're equal.

[1] with extra grep magic for xpimd.vty

I'll try to cook up an updated patch set tomorrow. Any thoughts and feedback much appreciated!

Thanks,
--Gabriel

**Comment by gsomlo** [comment]
> The more I think about it, the more I'd like to wrap flock around the whole quaggaboot.sh script call

Except if I do that, we get deadlock, since the daemon being launched prevents the lock from being released.

Also, I'm starting to wonder if each daemon's dependency on "zebra.vty" is somehow important.

At this point, I'd like to close this pull request, and will be submitting a new one once I come up with something cleaner...

**Comment by siliconja** [comment]
I'm following your comments and logic. I just don't have anything intelligent to add at the moment. Yes, preventing flock from hanging is important.

**Comment by gsomlo** [comment]
On Thu, Jan 12, 2017 at 08:39:43AM -0800, Jeff Ahrenholz wrote:
> I'm following your comments and logic. I just don't have anything
> intelligent to add at the moment. Yes, preventing flock from hanging
> is important.

Thanks, I appreciate knowing you're still interested!

I too care about reliable Quagga startup *a lot*: In my use case,
I need to start an unattended (Fedora) VM in which launches core-gui
auto-starts a simulation with some 70 BGP nodes and a few extra
containers running miscellaneous stuff. It has to work 100% perfectly
each time (VMWare creates and starts a bunch of instances of this VM
automatically, and having to go in and figure out if everything came
up OK is a 100% dealbreaker :)

I'll send a separate pull request with the systemd unit file one-liner
(TasksMax=infinity) because it's orthogonal to everything else, and
large simulations do bump into the default systemd fork branching
limit.

As for Quagga, I'm thinking hard of ways to do the right thing.

How about this: Since the old vtysh service *already* needs to know
which daemons should be up (since it has to wait for all their *.vty
files), wouldn't it make sense to make each daemon's start command
a no-op, and have a single designated service (vtysh, or if we remove
it, zebra) start everything that's listed and greppable in Quagga.conf
(just like currently done by bootvtysh() to find what to wait for) ?

Was this considered and rejected already, for reasons I'm about to
rediscover? :)

Thanks,
--Gabriel

**Comment by siliconja** [comment]
(Cool use case, BTW!)

I don't think we considered starting all daemons from vtysh startup, but yes that seems good. Still need to consider if you started the daemon but the vty didn't come up (crash etc.).

Some day in the future it would be nice to support both unified Quagga config and non-unified. Custom-built Quagga and the default distro one (deb/rpm).

**Pull request closed without merging**